README.md
# GePlu
==============================

 - Projeto criado com o objetivo de facilitar o processamento de uma enorme amostra de dados relacionados a precipitação.
 - A partir de registros de precipitação de uma estação pluviométrica, obtém a precipitação acumulada em diversos intervalos de tempo.
 - Busca realizar a tarefa da forma mais eficiente possível.
 
------------------------------


 


Releases.txt
Notas da Versão:

Beta 1.0:

- Versão inicial do programa, inclui as funcionalidades básicas
- Tempo médio de processamento -> arquivo utilizado com 55'245 linhas: 1 minuto e 55 segundos.

Beta 1.2:

- Adicionada a opção de selecionar o tipo de separador de colunas antes de processar o arquivo.
- Adicionada a opcao de ler/gerar um arquivo em formato .txt, além do formato .csv
- Removido temporariamente o dialogo de progresso.
- Notifica o usuário ao terminar a tarefa.
- Tempo médio de processamento -> arquivo utilizado com 55'250 linhas: 48 segundos.

Beta 1.3:

- Otimizado! (novas logica para o calculo)
- Tempo médio de processamento na Máquina de Teste -> arquivo utilizado com 55'250 linhas: 3 segundos.

======= Releases ========

Versão 1.0

- Novos Recursos adicionados:
	. Ao abrir um arquivo, é possível visualizar previamente como será a leitura do mesmo
	. Também é possível selecionar quais colunas serão utilizadas e o tipo da mesma.
- Adicionada uma Check Box para ignorar ou considerar a primeira linha do seu arquivo. Por padrão ela é considerada.
- Adicionado novamente um dialogo de progresso.

Versão 1.02

- Novos Recursos adicionados.
- Interface melhorada.
- Tempo médio de processamento na Máquina de Teste -> arquivo utilizado com 55'250 linhas: 5.5 segundos.

Versão 1.03

- Correção de bugs
- Tempo médio de processamento na Máquina de Teste -> arquivo utilizado com 55'250 linhas: 17 segundos.


classes/debug.py
from PySide2.QtWidgets import (QMainWindow, QApplication, QProgressDialog)
from datetime import timedelta, datetime

class Acumulados():
    ''' Tentativa de criar uma classe que representa todos os acumulados de uma estacao.'''

    def __init__(self, ChuvaAcumulada = [], dicChuva = {}, dicNivel = {}):
        self.Chuva = dicChuva #Dicionario contendo o horario e seus respectivos acumulados
        self.valuesIn = ChuvaAcumulada # cada inteiro representa minuto.
        self.Nivel = dicNivel #Dicionario contendo o horario e seus respectivos niveis do rio, se houver.
        self.isNivel = True if len(dicNivel) != 0 else False

    def getValues(self):
        return self.valuesIn

    def setData(self, Data):
        self.Chuva = Data

    def setDataHeight(self, Data, Ativar  = True):
        self.isNivel = Ativar
        self.Nivel = Data

    def setValues(self, Values):
        self.valuesIn = sorted(Values)

    def calcularAcumulado(self, parent):    
        ''' Calcula todos os acumulados do documento
        intervalos consiste em uma lista contendo o tempo dos acumulados solicitados, sao eles em: (1,4,24,48,96) horas'''
        intervalos = self.valuesIn # [60, 240, 1440, 2880, 5760] Minutos
        intervalosTime = [timedelta(minutes = x) for x in self.valuesIn]
        tamanho = len(intervalos) 
        dicionario = self.Chuva
        ArrayDatas = tuple(sorted(dicionario.keys()))

        # Constroi o Dialogo de Progresso
        progress = QProgressDialog("Gerando os valores de chuva acumulada...","Cancelar", 0, len(dicionario), parent = parent)
        progress.setMinimumDuration(0)
        progress.setWindowTitle("Progresso")
        progress.setModal(True)
        j = 0

        for pos in range(len(ArrayDatas)): #posicoes = [15/04/2018 10:00, 15/04/2018 10:15, ..]
            QApplication.processEvents()
            progress.setValue(j)
            if progress.wasCanceled():
                # Cancelado pelo usuario
                break
            j += 1

            currentKey = ArrayDatas[pos]
            count = 0
            posAcumulado = 0
            somatorio = 0.0
            limites = [currentKey - time for time in intervalosTime]
            while True:
                indice = pos + count
                if indice == -1:
                    dicionario[currentKey][posAcumulado + 2] = somatorio
                    posAcumulado += 1
                    if posAcumulado == tamanho:
                        break
                    else:
                        continue
                        
                lastKey = ArrayDatas[indice]
                if lastKey > limites[posAcumulado]:
                    # Do something
                    valores = dicionario[lastKey][:2]
                    if valores[0]:
                        somatorio += valores[1]
                    count -= 1
                else:
                    # Do something2
                    dicionario[currentKey][posAcumulado + 2] = somatorio
                    posAcumulado += 1
                    if posAcumulado == tamanho:
                        break
                
        progress.close()
        self.Chuva = dicionario

    def getChart(self):
        ''' retorna um lista, cada item da lista representa uma linha.
        Por sua vez, cada linha contem uma lista separada em: [data hora, acumulados, ..., nivel do rio]'''
        Universal = []
        linhaInicial = ["Data Hora", "Último Registro"]
        ini = "Chuva Acumulada em "
        for time in self.valuesIn:
            if time > 60*96:
                linhaInicial.append(ini + str(time//(60*24)) + " Dias")
            elif time >= 60:
                linhaInicial.append(ini + str(time//60) + " Hora(s)")
            else:
                linhaInicial.append(ini + str(time) + " Min")

        if self.isNivel:
            linhaInicial.append("Nivel do Rio")
            
        Universal.append(linhaInicial)

        #Formatando o texto data/hora de yyyy-mm-dd 00:00:00 para dd/mm/yyyy 00:00.
        for datetime1 in self.Chuva.keys():
            linha = []

            texto = datetime1.isoformat(" ")
            data, hora = texto.split(" ")
            data = data.split("-")
            data.reverse()
            data  = "/".join(data)
            hora = hora[:5]
            texto = data + " " + hora
            linha.append(texto)

            #Agora adicionando os valores dos acumulados a linha.
            first = True
            for acumulado in self.Chuva[datetime1][1:]:
                ac = 0
                if first:
                    ac = str(acumulado).replace('.',',')
                    first = False
                else:
                    ac = str(round(acumulado, 1)).replace('.',',')
                linha.append(ac)

            #Adicionando o respectivo nivel do rio, se houver.
            if self.isNivel:
                linha.append(self.Nivel[datetime1])

            #Adicionando a linha na lista Universal.
            Universal.append(linha)

        return Universal



classes/Generator.py
from PySide2.QtWidgets import (QMainWindow, QApplication, QProgressDialog)
from datetime import timedelta, datetime

class Acumulados():
    ''' Tentativa de criar uma classe que representa todos os acumulados de uma estacao.'''

    def __init__(self, ChuvaAcumulada = [], dicChuva = {}, dicNivel = {}):
        self.Chuva = dicChuva #Dicionario contendo o horario e seus respectivos acumulados
        self.valuesIn = ChuvaAcumulada # cada inteiro representa minuto.
        self.Nivel = dicNivel #Dicionario contendo o horario e seus respectivos niveis do rio, se houver.
        self.isNivel = True if len(dicNivel) != 0 else False

    def getValues(self):
        return self.valuesIn

    def setData(self, Data):
        self.Chuva = Data

    def setDataHeight(self, Data, Ativar  = True):
        self.isNivel = Ativar
        self.Nivel = Data

    def setValues(self, Values):
        self.valuesIn = sorted(Values)

    def calcularAcumulado(self, parent):    
        ''' Calcula todos os acumulados do documento
        intervalos consiste em uma lista contendo o tempo dos acumulados solicitados, sao eles em: (1,4,24,48,96) horas'''
        intervalos = self.valuesIn # [60, 240, 1440, 2880, 5760] Minutos
        intervalosTime = [timedelta(minutes = x) for x in self.valuesIn]
        tamanho = len(intervalos) 
        dicionario = self.Chuva
        ArrayDatas = tuple(sorted(dicionario.keys()))

        # Constroi o Dialogo de Progresso
        progress = QProgressDialog("Gerando os valores de chuva acumulada...","Cancelar", 0, len(dicionario), parent = parent)
        progress.setMinimumDuration(0)
        progress.setWindowTitle("Progresso")
        progress.setModal(True)
        j = 0

        for pos in range(len(ArrayDatas)): #posicoes = [15/04/2018 10:00, 15/04/2018 10:15, ..]
            QApplication.processEvents()
            progress.setValue(j)
            if progress.wasCanceled():
                # Cancelado pelo usuario
                break
            j += 1

            currentKey = ArrayDatas[pos]
            count = 0
            posAcumulado = 0
            somatorio = 0.0
            limites = [currentKey - time for time in intervalosTime]
            while True:
                indice = pos + count
                if indice == -1:
                    dicionario[currentKey][posAcumulado + 2] = somatorio
                    posAcumulado += 1
                    if posAcumulado == tamanho:
                        break
                    else:
                        continue
                        
                lastKey = ArrayDatas[indice]
                if lastKey > limites[posAcumulado]:
                    # Do something
                    valores = dicionario[lastKey][:2]
                    if valores[0]:
                        somatorio += valores[1]
                    count -= 1
                else:
                    # Do something2
                    dicionario[currentKey][posAcumulado + 2] = somatorio
                    posAcumulado += 1
                    if posAcumulado == tamanho:
                        break
                
        progress.close()
        self.Chuva = dicionario

    def getChart(self):
        ''' retorna um lista, cada item da lista representa uma linha.
        Por sua vez, cada linha contem uma lista separada em: [data hora, acumulados, ..., nivel do rio]'''
        Universal = []
        linhaInicial = ["Data Hora", "Último Registro"]
        ini = "Chuva Acumulada em "
        for time in self.valuesIn:
            if time > 60*96:
                linhaInicial.append(ini + str(time//(60*24)) + " Dias")
            elif time >= 60:
                linhaInicial.append(ini + str(time//60) + " Hora(s)")
            else:
                linhaInicial.append(ini + str(time) + " Min")

        if self.isNivel:
            linhaInicial.append("Nivel do Rio")
            
        Universal.append(linhaInicial)

        #Formatando o texto data/hora de yyyy-mm-dd 00:00:00 para dd/mm/yyyy 00:00.
        formato = "%d/%m/%Y %H:%M"
        for datetime1 in self.Chuva.keys():
            linha = []
            linha.append(datetime1.strftime(formato))

            #Agora adicionando os valores dos acumulados a linha.
            first = True
            for acumulado in self.Chuva[datetime1][1:]:
                ac = 0
                if first:
                    ac = str(acumulado).replace('.',',')
                    first = False
                else:
                    ac = str(round(acumulado, 1)).replace('.',',')
                linha.append(ac)

            #Adicionando o respectivo nivel do rio, se houver.
            if self.isNivel:
                linha.append(self.Nivel[datetime1])

            #Adicionando a linha na lista Universal.
            Universal.append(linha)

        return Universal



classes/functions.py
from datetime import timedelta, datetime
from PySide2.QtWidgets import (QMainWindow, QMessageBox, QDialog)
import csv


def toDict(self, archive, colunas):
    objeto = self.Gerador
    formatoData = self.FormatoData.currentText()
    dicChuva = {}
    dicNivel = {}
    chaves = colunas.keys()
    itens = len(objeto.getValues())
    AreTogether = "Data & Hora" in chaves
    isNivel = "Nível do Rio" in chaves

    #Testes de caso e manuseio de erros.
    if not AreTogether:
        if not "Data" in chaves or not "Hora" in chaves:
            x = QMessageBox(QMessageBox.Warning, "Erro de Seleçao",  'A coluna correspondente a "Data" ou "Hora" não foi selecionada', parent = self)
            x.addButton(QMessageBox.Ok)
            x.setInformativeText("Por favor, verifique a tabela e selecione a coluna correta.")
            x.exec()
            raise KeyError

    if not "Chuva Medida" in chaves:
        x = QMessageBox(QMessageBox.Warning, "Erro de Seleçao",  'A coluna "Chuva Medida" não foi selecionada', parent = self)
        x.addButton(QMessageBox.Ok)
        x.setInformativeText("Por favor, verifique a tabela e selecione a coluna correta.")
        x.exec()
        raise KeyError

    for row in archive:
        try:
            # Formata o texto de DataHora e convertem em um tipo datetime.
            time, date = '', ''
            if not AreTogether:
                time = row[colunas["Hora"]]
                date = row[colunas["Data"]]
            else:
                date, time = row[colunas["Data & Hora"]].split(" ")

            if formatoData == "DD/MM/AAAA":
                date = date.split("/") if "/" in date else date.split('-')
                date.reverse()
                for num in range(len(date)):
                    if len(date[num]) == 1:
                        date[num] = '0'+date[num]
                date = "-".join(date)
            else:
                date = date.replace("/","-") 
            if '.' in time:
                x = str.index(time, '.')
                time = time[:x]
                
            timedate = datetime.fromisoformat(date + ' ' + time)

        except ValueError:
            x = QMessageBox(QMessageBox.Critical, "Erro de Formatação",  "O formato especificado para Data ou Hora é inválido.", parent = self)
            x.addButton(QMessageBox.Ok)
            x.setInformativeText("Por favor, verifique o formato da data ou o conteúdo da coluna.")
            x.exec()
            raise ValueError
        # Valor do Acumulado
        try:
            item = row[colunas["Chuva Medida"]].replace(",", ".")
            teste = False
            if '.' in item or item.isnumeric():
                item = float(item)
                teste = True
                
            dicChuva[timedate] = [teste, item] + [0]*itens

            #Valor do Nivel do Rio, se houver.
            if isNivel:
                dicNivel[timedate] = row[colunas["Nível do Rio"]]
        except ValueError:
            x = QMessageBox(QMessageBox.Critical, "Erro de Formatação",  "Não foi possível identificar o valor registrado", parent = self)
            x.addButton(QMessageBox.Ok)
            x.setInformativeText('Por favor, verifique o conteúdo da coluna "Chuva Medida".')
            x.exec()
            raise ValueError

        except IndexError:
            x = QMessageBox(QMessageBox.Critical, "Erro de Seleção",  "Não foi possível identificar o valor registrado", parent = self)
            x.addButton(QMessageBox.Ok)
            x.setInformativeText('A Coluna Selecionada para "Chuva Medida" está vazia ou não existe.')
            x.exec()
            raise IndexError

    objeto.setData(dicChuva)
    if len(dicNivel) > 0:
        objeto.setDataHeight(dicNivel)
    return None


def openFile(parent):
    ''' Abre o arquivo e adiciona os seus dados a dois dicionarios.
    o arquivo .csv deve conter as colunas representando a data/hora, o ult.acumulado e o nivel do rio (opcional)
    cada dicionario contem os dados de Chuva e Nivel do rio, respectivamente.'''
    Archive = []
    Ignore = parent.ignore.isChecked()
    try:
        with open(parent.path.text(), newline = '') as csvfile:
            newfile = csv.reader(csvfile, delimiter = parent.separador)
            for row in newfile:
                # Primeira linha.
                if Ignore:
                    Ignore = False
                    continue
                if len(row[0]) > 0:
                    Archive.append(row)

    except FileNotFoundError as Err:
        x = QMessageBox(QMessageBox.Critical, "Erro de Acesso", "Não foi possível abrir o arquivo solicitado.", buttons = QMessageBox.Ok, parent = parent)
        x.setInformativeText("Não existe tal arquivo ou diretório.")
        x.setDetailedText(str(Err))
        x.exec()
        raise FileNotFoundError

    return Archive

def saveFile(parent, File, Delimiter = ';'):

    try:
        with open(parent.folder.text(), mode = 'w', newline = '') as csvfile:
            arquivo = csv.writer(csvfile, delimiter = Delimiter, quoting = csv.QUOTE_MINIMAL)
            for row in File:
                arquivo.writerow(row)
    except PermissionError as Err:
        x = QMessageBox(QMessageBox.Critical, "Erro de Acesso", "Não foi possível salvar seu arquivo.", buttons = QMessageBox.Ok, parent = parent)
        x.setInformativeText("O arquivo que voce está tentando sobrescrever já está aberto em outro programa.")
        x.setDetailedText(str(Err))
        x.exec()
        raise PermissionError


main.py
# Arquivo da Interface GUI do programa.
# Python 3.7.7

from PySide2.QtGui import (QIcon)
from PySide2.QtCore import (QUrl, QObject)
from PySide2.QtWidgets import (QMainWindow, QApplication, QLineEdit, QVBoxLayout, QWidget, QHBoxLayout,
QBoxLayout, QPushButton, QGridLayout, QRadioButton, QGroupBox, QLabel, QTableWidget, QComboBox, QFileDialog,
QTableWidgetItem, QProgressDialog, QMessageBox, QCheckBox, QDialog, QInputDialog)
import classes.Generator as Ac
import classes.functions as Fc
import sys, csv, time, os, ctypes

class MainWindow(QMainWindow):

    def __init__(self):
        QMainWindow.__init__(self)
        # Variaveis
        self.Gerador = Ac.Acumulados([60, 60*4, 60*24, 60*48, 60*96])
        self.separador = ";"

        #Janela Principal
        widget = QWidget()
        self.setCentralWidget(widget)

        # Inicializa os Widgets
        self.folder = QLineEdit("Salvar Como...")
        self.path = QLineEdit("Abrir arquivo...")

        buttonOpen = QPushButton('Abrir')
        buttonSave = QPushButton("Destino")
        Processar = QPushButton('Executar')
        Ajuda = QPushButton('Informações')

        groupBox2 = QGroupBox("Delimitador")
        self.delimitador1 = QRadioButton("Ponto-Vírgula")
        self.delimitador2 = QRadioButton("Vírgula")
        self.delimitador3 = QRadioButton("Ponto")

        text3 = QPushButton("Configurar Intervalos")
        text2 = QLabel("Formato da Data")
        self.FormatoData = QComboBox()
        self.FormatoData.addItems(["DD/MM/AAAA",'AAAA/MM/DD'])
        checkGroup = QGroupBox("Mais opções")

        text = QLabel("Por favor, selecione na tabela abaixo as colunas a utilizar:")
        self.ignore = QRadioButton("Ignorar Primeira Linha do Arquivo")

        self.Tabela = QTableWidget(15,15)
        self.startTable()

        # Layouts
        MainLayout = QVBoxLayout()

        Gridlayout = QGridLayout()
        Gridlayout.addWidget(self.path, 0, 0)
        Gridlayout.addWidget(self.folder, 1, 0)
        Gridlayout.addWidget(buttonOpen, 0, 1)
        Gridlayout.addWidget(buttonSave, 1, 1)
        Gridlayout.addWidget(Processar, 0, 3)
        Gridlayout.addWidget(Ajuda, 1, 3)
        Gridlayout.setColumnStretch(0, 2)
        Gridlayout.setColumnStretch(3, 1)
        Gridlayout.setColumnMinimumWidth(2, 20)

        SecondLayout = QHBoxLayout()
        SecondLayout.addWidget(groupBox2)
        SecondLayout.addSpacing(70)
        SecondLayout.addWidget(checkGroup)
        #
        SepLayout = QVBoxLayout()
        SepLayout.addWidget(self.delimitador1)
        SepLayout.addWidget(self.delimitador2)
        SepLayout.addWidget(self.delimitador3)
        #
        OptionsLayout = QVBoxLayout()
        OptionsLayout.addWidget(text3)
        OptionsLayout.addWidget(text2)
        OptionsLayout.addWidget(self.FormatoData)

        ThirdLayout = QVBoxLayout()
        ThirdLayout.addWidget(self.ignore)
        ThirdLayout.addWidget(text)

        MainLayout.addLayout(Gridlayout)
        MainLayout.addLayout(SecondLayout)
        MainLayout.addLayout(ThirdLayout)
        MainLayout.addWidget(self.Tabela)

        # Coloca o Layout principal na Janela
        widget.setLayout(MainLayout)

        # Comandos dos Widgets e edicoes.
        groupBox2.setLayout(SepLayout)
        self.delimitador1.setChecked(True)
        self.folder.setReadOnly(True)
        self.path.setReadOnly(True)
        checkGroup.setLayout(OptionsLayout)

        buttonOpen.clicked.connect(self.searchFile)
        buttonSave.clicked.connect(self.getNewFile)
        self.delimitador1.clicked.connect(self.updateDelimiter)
        self.delimitador2.clicked.connect(self.updateDelimiter)
        self.delimitador3.clicked.connect(self.updateDelimiter)
        self.ignore.clicked.connect(self.updateIgnoreButton)
        Ajuda.clicked.connect(self.help)
        Processar.clicked.connect(self.taskStart)
        text3.clicked.connect(self.openSubWindow)

        # Propriedades da janela principal
        myappid = 'GePlu.release1_03' # arbitrary string
        ctypes.windll.shell32.SetCurrentProcessExplicitAppUserModelID(myappid)
        self.setWindowIcon(QIcon('icon6.ico'))
        self.setFixedSize(640,480)
        self.setWindowTitle("GePlu")

    def openSubWindow(self):

        dialog = MyDialog(self.Gerador)
        dialog.show()
        dialog.exec_()

    def taskStart(self):
        ''' Inicia a execucao do programa, se houver algum erro durante o processo
        notifica o usuario e deixa a funcao imediatamente, devido a isso eu escrevi dois linhas
        para garantir que os dados no objeto Gerador estejam resetados e o objeto esteja pronto para
        a proxima interacao'''
        if self.folder.isModified() and self.path.isModified(): # o usuario entrou com os caminhos (paths)?
            # Garante que os dados são novos, ou seja, reseta se ja houve uma interacao antes.
            self.Gerador.setData({})
            self.Gerador.setDataHeight({},False)
            # Verifica os nomes das colunas na tabela do programa e armazena em uma variavel
            colunas = {}
            for col in range(self.Tabela.columnCount()):
                colunas[self.Tabela.cellWidget(0,col).currentText()] = col
            # Comeca a cronometrar
            start = time.time()
            # Abre o arquivo e armazena em uma variavel
            rawText = Fc.openFile(self)
            # Pega o arquivo e converte os dados de data/hora em datetime, e valores de registro (ultima chuva) em floats, se possivel.
            # Depois coloca o resultado dentro do objeto Gerador
            Fc.toDict(self, rawText, colunas)
            rawText = None
            # Calcula todos os acumulados em x,y,z horas (especificado pelo usuario) de todos os horarios
            self.Gerador.calcularAcumulado(self)
            # Poe o resultado em uma matriz e retorna ela
            Result = self.Gerador.getChart()
            # Abre um novo arquivo csv e escreve os dados da matriz nesse arquivo, por fim salva.
            Fc.saveFile(self, Result)
            # Fim do processo
            end = time.time()
            # Notifica o usuario
            QMessageBox.information(self, "Notificação", "Tarefa realizada com sucesso!\nTempo de execução: {} s.".format(round(end-start, 2)))
            # Reseta a tabela e os endereço do arquivo aberto e onde salvar, na janela principal.
            self.resetProgram()
        else:
            QMessageBox.warning(self, "Notificação", "Houve um erro no arquivo ou diretório especificado.\nPor favor, selecione um caminho válido.")


    def help(self):
        x = "Caso tenha alguma dúvida, abra o arquivo de texto 'Leia-me' presente\nna pasta do programa ou entre em contato\n\nVersão 1.03\nCriado por Lucas da Silva Menezes\nContato: lucasmenezes4502@gmail.com"
        msgBox = QMessageBox.information(self, "Informação", x)

    def resetProgram(self):
        self.Tabela.clearContents()
        self.startTable()
        self.folder.setText("Salvar Como...")
        self.path.setText("Selecionar o arquivo...")
        self.folder.setModified(False)
        self.folder.setModified(False)

    def updateIgnoreButton(self):
        if self.path.isModified():
            self.updateTable()

    def updateDelimiter(self):
        separadores = {"Ponto-Vírgula": ';', "Vírgula":",","Ponto":"."}
        for x in [self.delimitador1, self.delimitador3, self.delimitador2]:
            if x.isChecked():
                self.separador = separadores[x.text()]
                break
        if self.path.isModified():
            self.updateTable()

    def startTable(self):
        for col in range(self.Tabela.columnCount()):
            combo = QComboBox()
            combo.addItems(["Selecionar","Data & Hora","Chuva Medida","Data","Hora","Nível do Rio","Ignorar"])
            self.Tabela.setCellWidget(0, col, combo)

    def updateTable(self):
        textos = [0]*15
        for col in range(15):
            textos[col] = self.Tabela.cellWidget(0, col).currentText()

        self.Tabela.clearContents()
        self.startTable()

        for col in range(15):
            self.Tabela.cellWidget(0, col).setCurrentText(textos[col])

        resultado = self.ignore.isChecked()
        try:
            with open(self.path.text(), newline = '') as csvfile:
                file = csv.reader(csvfile, delimiter = self.separador)
                n = 0
                for row in file:
                    if n == 0 and resultado:
                        resultado = False
                        continue
                    for col in range(len(row)):
                        if col < self.Tabela.columnCount():
                            self.Tabela.setItem(n + 1, col, QTableWidgetItem(row[col]))
                    n += 1
                    if n == 14:
                        break
        except FileNotFoundError as Err:
            x = QMessageBox(QMessageBox.Critical, "Erro de Acesso",  "Não foi possível abrir o arquivo selecionado.", parent = self)
            x.addButton(QMessageBox.Ok)
            x.setInformativeText("Não existe tal arquivo ou diretório.")
            x.setDetailedText(str(Err))
            x.exec()
            raise PermissionError


    def searchFile(self):
        address, x = QFileDialog.getOpenFileName(self, "Selecione um arquivo", filter = "Text files (*.txt *.csv)")
        if len(address) > 0:
            self.path.setText(address)
            self.path.setModified(True)
            self.updateTable()

    def getNewFile(self):
        address, x = QFileDialog.getSaveFileName(self, "Salvar como", filter = "Text files (*.csv);; Text files (*.txt)")
        if len(address) > 0:
            self.folder.setText(address)
            self.folder.setModified(True)

class MyDialog(QDialog):

    def __init__(self, gerador):
        QDialog.__init__(self)
        self.setWindowTitle("Configuração")
        self.setFixedSize(200,200)
        self.setWindowIcon(QIcon('icon6.ico'))
        self.setModal(True)
        self.intervalos = [60, 60*2, 60*3, 60*4, 60*12, 60*24, 60*48, 60*72, 60*96, 60*24*30]

        button1 = QPushButton("Cancelar")
        button1.clicked.connect(self.closeDialog)
        button2 = QPushButton("Confirmar")
        button2.clicked.connect(lambda: self.saveAndClose(gerador))

        horarios = ["1 Hora","2 Horas", "3 Horas", "4 Horas","12 Horas",
         "24 Horas", "48 Horas", "72 horas", "96 horas", "30 Dias"]
        self.checkboxes = [0]*10
        for i in range(10):
            self.checkboxes[i] = QCheckBox(horarios[i])

        grupo = QGroupBox("Incluir Precipitação Acumulada Em")
        grid = QGridLayout()
        for x in range(10):
            if x < 5:
                grid.addWidget(self.checkboxes[x], x , 0)
            else:
                grid.addWidget(self.checkboxes[x], x - 5, 1)
        grupo.setLayout(grid)
        self.startCheckboxes(gerador)

        sublayout = QHBoxLayout()
        sublayout.addWidget(button1)
        sublayout.addWidget(button2)

        layout = QVBoxLayout()
        layout.addWidget(grupo)
        layout.addLayout(sublayout)
        self.setLayout(layout)

    def startCheckboxes(self, gerador):
        for x in range(10):
            self.checkboxes[x].setChecked(self.intervalos[x] in gerador.getValues())

    def saveAndClose(self, gerador):
        connections = []
        for x in range(10):
            if self.checkboxes[x].isChecked():
                connections.append(self.intervalos[x])

        gerador.setValues(connections)
        self.close()

    def closeDialog(self):
        self.close()


if __name__ == "__main__":
    app = QApplication(sys.argv)
    # Cria e Inicia a janela principal
    window = MainWindow()
    window.show()
    # Loop principal
    sys.exit(app.exec_())
